solutions/gold/usaco-1235.mdx

```diff
@@ -0,0 +1,171 @@
+// returns all restrictions if valid
+// otherwise, returns an empty array
+vector<pair<int, int>> check(int k) {
+	vector<pair<int, int>> lr(n + 1);
+	for (int i = 1; i <= n; i++) {
+		// all ancestors of i are restricted to [l_i - k, r_i + k]
+		lr[i] = {L[i] - k, R[i] + k};
+	}
+
+	// this works because p_i < i
+	for (int i = n; i > 0; i--) {
+		// check for degenerate ranges
+		if (lr[i].first > lr[i].second) { return {}; }
+		if (p[i] && (lr[i].first > R[p[i]] || lr[i].second < L[p[i]])) {
+			return {};
+		}
+		// propagate restrictions to parent
+		lr[p[i]] = {
+			max(lr[p[i]].first, lr[i].first),
+			min(lr[p[i]].second, lr[i].second)
+		}
+	}
+
+	return {};
+}
+```
+
+</CPPSection>
+</LanguageSection>
+</Spoiler>
+<Spoiler title="B = 1">
+Now that we have our minimum imbalance, all that's left to do is construct a solution! We can essentially reverse our process for checking: nodes will be more restricted than their descendants, with the root being the most restricted and the leaves being the least.
+
+Thus, we can build our solution top-down: starting at node 1, we can pick any value satisfying all restrictions found from the `check` function (in the solution, we simply pick the left endpoint of the range of values for simplicity). Then, if this value is $x$, all descendants are restricted by $[x - k, x + k]$ (in addition to their own restrictions from `check`).
+
+Here's the relevant snippet:
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+void solve(int k) {
+	// call check to get restrictions
+	auto lr = check(k);
+	for (int i = 0; i <= n; i++) {
+		// propagate `check` restrictions to ancestors
+		L[p[i]] = max(L[p[i]], lr[i].first);
+		R[p[i]] = min(R[p[i]], lr[i].second);
+	}
+
+	// l and r: restrictions
+	// on i's descendants rather than ancestors
+	vector<int> l(n + 1, 0);
+	vector<int> r(n + 1, K);
+
+	for (int i = 1; i <= n; i++) {
+		// pick s[i] to be left endpoint of valid range
+		// (intersection of [L_i, R_i] and [l_p[i], r_p[i]])
+		s[i] = max(L[i], l[p[i]]);
+		// restrict children of i to range [s_i - k, s_i + k]
+		l[i] = max(l[p[i]], s[i] - k);
+		r[i] = min(r[p[i]], s[i] + k);
+	}
+}
+```
+
+</CPPSection>
+</LanguageSection>
+</Spoiler>
+
+<Spoiler title="Full Code">
+And here's the full code:
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <vector>
+using namespace std;
+
+const int N = 1e5 + 1;
+const int K = 1e9;
+
+int n, p[N], L[N], R[N], s[N];
+
+// for a more detailed explanation of check and solve see above
+vector<pair<int, int>> check(int k) {
+	vector<pair<int, int>> lr(n + 1);
+	for (int i = 1; i <= n; i++) { lr[i] = {L[i] - k, R[i] + k}; }
+
+	for (int i = n; i > 0; i--) {
+		if (lr[i].first > lr[i].second) { return {}; }
+		if (p[i] && (lr[i].first > R[p[i]] || lr[i].second < L[p[i]])) {
+			return {};
+		}
+		lr[p[i]] = {max(lr[p[i]].first, lr[i].first),
+		            min(lr[p[i]].second, lr[i].second)};
+	}
+
+	return lr;
+}
+
+void solve(int k) {
```

Comment:
could you put this function right in main? if you're worried about the snippet you can just cut out the function and unindent everything up there one level